python/ray/tune/schedulers/trial_scheduler_v2.py

```diff
@@ -0,0 +1,237 @@
+        that react to metrics with their own `metric` and `mode` parameters.
+
+        Args:
+            metric (str): Metric to optimize
+            mode (str): One of ["min", "max"]. Direction to optimize.
+        """
+        if self._metric and metric:
+            return False
+        if metric:
+            self._metric = metric
+
+        if self._metric is None:
+            # Per default, use anonymous metric
+            self._metric = DEFAULT_METRIC
+
+        return True
+
+    def on_trial_add(self, trial: Trial):
+        """Called when a new trial is added to the trial runner."""
+
+        raise NotImplementedError
+
+    def on_trial_error(self, trial: Trial):
+        """Notification for the error of trial.
+
+        This will only be called when the trial is in the RUNNING state."""
+
+        raise NotImplementedError
+
+    def on_trial_result(self, trial: Trial, result: Dict) -> str:
+        """Called on each intermediate result returned by a trial.
+
+        At this point, the trial scheduler can make a decision by returning
+        one of CONTINUE, PAUSE, and STOP. This will only be called when the
+        trial is in the RUNNING state."""
+
+        raise NotImplementedError
+
+    def on_trial_complete(self, trial: Trial, result: Dict):
+        """Notification for the completion of trial.
+
+        This will only be called when the trial is in the RUNNING state and
+        either completes naturally or by manual termination."""
+
+        raise NotImplementedError
+
+    def on_trial_remove(self, trial: Trial):
+        """Called to remove trial.
+
+        This is called when the trial is in PAUSED or PENDING state. Otherwise,
+        call `on_trial_complete`."""
+
+        raise NotImplementedError
+
+    def choose_trial_to_run(self,
+                            has_resources_for_trial_cb: Callable[[Trial], bool]
+                            ) -> Optional[Trial]:
+        """Called to choose a new trial to run.
+
+        This should return one of the trials in trial_runner that is in
+        the PENDING or PAUSED state. This function must be idempotent.
+
+        If no trial is ready, return None."""
+
+        raise NotImplementedError
+
+    def fetch_decisions(self) -> Dict[str, List[Trial]]:
+        """Outputs decisions for Tune loop to act upon.
+
+        This is usually called after `on_trial_result` (primary use).
+        Sometimes, it's also called in an iterative fashion during other
+        triggers into scheduler. For example, `on_trial_complete` may
+        generate some new scheduler decisions to be executed;
+        `fetch_decision` is called and during the execution of these decisions,
+         new triggers into scheduler may be called again (secondary use).
+
+        The result contains a list of trials of each kind of decisions.
+        TrialRunner executes these decisions following certain priorities. See
+        comments under TrialSchedulerV2.
+        """
```

Comment:
I don't totally know the context behind this, but I'm not sure if this is a good idea.

It feels to me that the state here is very implicit. How do I know if I'm calling 'fetch_decision" at the right time?

Reply:
I thought this is what we talked about on Wednesday when we said we want to decouple `decision` from a single trial, to make it more semantically clear. I actually have no preference about this. Alternatively, we could just have `on_trial_result` and `on_trial_xxx` all return decisions to lists of trials mapping.